C++: Use pyrameterized modules for TaintTracking

jbj opened the pull request:
This is a minimal PR to demonstrate how taint tracking can be pyrameterized such that each copy of the taint-tracking configuration class is byte-for-byte identical. It also switches the C++ interface from using `TaintTracking::Configuration2` to using `TaintTracking2::Configuration` like C# does it.

The commits should make sense individually: the first commit rearranges the data flow library, and the second commit makes the actual change to C++ taint tracking.

I didn't go as far as unifying the taint-tracking configuration classes across languages. I can follow up with language-specific PRs for that if we merge this PR.

hvitved commented on cpp/ql/src/semmle/code/cpp/dataflow/TaintTracking2.qll:
```diff
@@ -0,0 +1,13 @@
+ * 100` since we consider a single bit of information to be too little.
+ */
+module TaintTracking2 {
+  import semmle.code.cpp.dataflow.internal.TaintTrackingUtil
```
Shouldn't this import happen in `dataflow2/TaintTrackingImpl.qll` (and same for `TaintTracking.qll`)?

jbj replied:
It probably should. That would be consistent with how we do it for `DataFlow`. I'll change that.

jbj commented:
I simplified this PR vastly after talking to @aschackmull. I dropped the commit that changed the data-flow libraries, so this PR now only changes the C/C++ taint tracking library. That means fewer code changes and fewer confusing indirections.

hvitved commented on cpp/ql/src/semmle/code/cpp/dataflow/internal/tainttracking2/TaintTrackingParameter.qll:
```diff
@@ -0,0 +1,2 @@
+import semmle.code.cpp.dataflow.DataFlow as Private
```
Should be `dataflow2`, right?

jbj replied:
Thanks for spotting!

jbj replied:
Fixed.

jbj commented:
Fixing the `DataFlow2` bug that Tom found revealed that `TaintTrackingParameter.qll` has to do a little more work: it has to rename `DataFlow2` to `DataFlow` in the scope of `tainttracking2/TaintTrackingImpl.qll`. I added that in  commit no. 3.

@aschackmull, are you happy enough with these changes that I should apply the same restructuring to Java (and make the taint tracking implementation files identical across languages)?

aschackmull commented:
> @aschackmull, are you happy enough with these changes that I should apply the same restructuring to Java (and make the taint tracking implementation files identical across languages)?

Yes.

aschackmull commented on java/ql/src/semmle/code/java/dataflow/TaintTracking.qll:
```diff
@@ -1,804 +1,24 @@
+ *
+ * We define _taint propagation_ informally to mean that a substantial part of
+ * the information from the source is preserved at the sink. For example, taint
+ * propagates from `x` to `x + 100`, but it does not propagate from `x` to `x >
```
This does not hold for java.

aschackmull commented on java/ql/src/semmle/code/java/dataflow/TaintTracking2.qll:
```diff
@@ -0,0 +1,12 @@
+ *
+ * We define _taint propagation_ informally to mean that a substantial part of
+ * the information from the source is preserved at the sink. For example, taint
+ * propagates from `x` to `x + 100`, but it does not propagate from `x` to `x >
```
Same.

jbj commented:
I now also made the change for C# and added change notes. I rebased the PR to avoid a change note conflict.

hvitved commented on csharp/ql/src/semmle/code/csharp/dataflow/internal/tainttracking1/TaintTrackingParameter.qll:
```diff
@@ -0,0 +1,5 @@
+
+module Private {
+  import semmle.code.csharp.dataflow.DataFlow::DataFlow as DataFlow
+}
```
I would prefer
```
import semmle.code.csharp.dataflow.internal.TaintTrackingPublic as Public

module Private {
  import semmle.code.csharp.dataflow.DataFlow::DataFlow as DataFlow
  import semmle.code.csharp.dataflow.internal.TaintTrackingPrivate
}
```
and get rid of `TaintTrackingUtil.qll`

jbj replied:
Done
